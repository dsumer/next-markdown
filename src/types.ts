export type NextMdConfig = {
  /**
   * The place where to find your markdown files and folders.
   *
   * This is mandatory.
   */
  pathToContent: string;

  /**
   * The place where your markdown files are stored.
   *
   * - if empty / undefined, your markdown files are considered to be in your current project at path `pathToContent`.
   * - if specified, nextmd will pull or clone the repository and look for markdown files at path `pathToContent` in that repo.
   */
  contentGitRemoteUrl?: string;
};

export type YAMLFrontMatter = { [key: string]: any };
