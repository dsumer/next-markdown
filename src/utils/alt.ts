/**
 * If the alt is a JSON, the JSON key/value will be transfered to `node.properties`.
 *
 * For example, this markdown:
 *
 * ```md
 * ![{"alt":"20211029 - edit bot demo","poster":"https://frouo.com/poster.jpg"}](https://frouo.com/video.mp4)
 * ```
 *
 * will generate:
 *
 * ```html
 * <video src="https://frouo.com/video.mp4" alt="20211029 - edit bot demo" width="100%" controls poster="https://frouo.com/poster.jpg"></video>
 * ```
 */
export const extractDataFromAlt = (alt: string) => {
  try {
    return JSON.parse(alt) as { [key: string]: any };
  } catch (error) {
    return alt;
  }
};