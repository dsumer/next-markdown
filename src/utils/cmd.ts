import { exec } from 'child_process';

export const cmd = (commandLine: string) => {
  return new Promise<string>((resolve, reject) => {
    exec(commandLine, (error, stdout, stderr) => {
      if (error) {
        reject(error);
      }
      resolve(stdout || stderr);
    });
  });
};
