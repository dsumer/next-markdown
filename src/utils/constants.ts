import { join } from 'path';

export const pathToLocalGitRepo = join(process.cwd(), '.git/next-md/');
export const pathToNextmdLastUpdate = join(process.cwd(), '.git/next-md-last-update');
export const pathToNextmdBranch = join(process.cwd(), '.git/next-md-branch');
