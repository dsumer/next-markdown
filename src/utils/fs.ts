import fs from 'fs';
import { join, parse, sep } from 'path';
import { File, TreeObject } from '../types';

export const pathToContent = (userPathToContent: string, remote: boolean) =>
  join(remote ? './.git/next-md' : './', userPathToContent);

export const exclude = (object: TreeObject) => {
  if (object.type === 'file' && object.name.endsWith('.md') === false && object.name.endsWith('.mdx') === false) {
    return true;
  }

  if (object.name === '.git') {
    return true;
  }

  if (object.type === 'dir' && object.children.length === 0) {
    return true;
  }

  return false;
};

export const treeSync = (path: string, absoluteToRelative: (path: string) => string) => {
  const res = fs
    .readdirSync(path)
    .map((e) => ({ name: e, dir: path }))
    .map(
      (
        (fileSystem: typeof fs) =>
        (e): TreeObject | null => {
          const _path = join(e.dir, e.name);
          const isDir = fileSystem.lstatSync(_path).isDirectory();
          if (e.name === '.git') {
            return null; // no need to inspect recursively this
          } else {
            return isDir
              ? {
                  type: 'dir',
                  name: e.name,
                  path: absoluteToRelative(_path),
                  children: treeSync(_path, absoluteToRelative),
                }
              : {
                  type: 'file',
                  name: e.name,
                  path: absoluteToRelative(_path),
                };
          }
        }
      )(fs),
    )
    .flatMap((f) => (f ? [f] : []))
    .filter((e) => exclude(e) === false);

  return res;
};

export const flatFiles = (tree: TreeObject[]): File[] => {
  const flatRecursively = (object: TreeObject): TreeObject[] =>
    object.type === 'dir' ? object.children.flatMap(flatRecursively) : [object];

  return tree.flatMap(flatRecursively) as File[];
};

export const generateNextmd = (relativeFilePath: string) => {
  return relativeFilePath
    .split(sep)
    .map((e) => e.replace(/^_(\s+)?/, '')) // replace the first occurence of "_" at the beginning of the string with ""
    .map((e) => e.replace(/^\[(.*?)\](\s+)?_?/, '')) // replace the first occurence of "[whatever-you-want]+whitespaces(if-any)" at the beginning of string with ""
    .map((e, index, arr) => {
      if (index === arr.length - 1) {
        const lastElement = e;
        const parsed = parse(lastElement);
        if (parsed.name === 'index') {
          return undefined;
        } else {
          return parsed.name;
        }
      } else {
        return e;
      }
    })
    .map((e) => e?.trim())
    .flatMap((e) => (e ? [e] : []));
};

export const getPostFilesFromNextmd = async (files: File[], nextmd: string[]) =>
  files.filter((e) => parse(e.path).dir === nextmd.join('/'));

export const isMDX = (filePath: string) => filePath.endsWith('.mdx');

export const readFileSyncUTF8 = (filePath: string) => fs.readFileSync(filePath).toString('utf-8');

export const isDraft = (relativeFilePath: string) =>
  relativeFilePath.split(sep).find((e) => e.startsWith('_')) !== undefined;
