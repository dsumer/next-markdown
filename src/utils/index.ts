export * from './alt';
export * from './fs';
export * from './git';
export * from './markdown';
export * from './logger';
