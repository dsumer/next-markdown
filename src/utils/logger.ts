export const consoleLogNextmd = (...args: (string | undefined | null)[]) => {
  if (process.env.NODE_ENV === 'test') {
    return;
  }

  args.unshift('[nextmd]');
  console.log.apply(this, args); // tslint:disable-line:no-console
};
