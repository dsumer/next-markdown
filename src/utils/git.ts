import fs from 'fs';
import { cmd } from './cmd';
import { treeSync } from './fs';
import { consoleLogNextmd } from './logger';

const pathToNextmdLastUpdate = '.git/next-md-last-update';
const pathToNextmdBranch = '.git/next-md-branch';

type GitRepo = {
  remoteUrl: string;
  branch: string;
};

/**
 * @param pathToContent Absolute path to content
 */
export const treeContentRepo = async (
  pathToContent: string,
  absoluteToRelative: (path: string) => string,
  debug: boolean,
  gitRepo?: GitRepo,
) => {
  if (gitRepo) {
    const { remoteUrl, branch } = gitRepo;
    if (!remoteUrl || !branch) {
      throw Error('[nextmd] You must specify both git remote URL and branch when using `contentGitRepo`');
    }

    /**
     * @returns The number of seconds since the last update of the git remote repo.
     */
    const elapsedSecondsSinceLastUpdate = () => {
      const lastRepoUpdateTxt = fs.readFileSync(pathToNextmdLastUpdate).toString('utf-8').trim();
      const lastUpdateMillis = parseInt(lastRepoUpdateTxt, 10);

      return Date.now() - lastUpdateMillis;
    };

    /**
     *
     * @returns The branche name of the last git clone
     */
    const branchLastUpdate = () => fs.readFileSync(pathToNextmdBranch).toString('utf-8').trim();

    /**
     * Mechanism to avoid pulling the repo when `getStaticPaths` & `getStaticProps` is called.
     *
     * This ensures repo content is the same in `getStaticPaths` & `getStaticProps`.
     */
    const shouldUpdateGitRepo = () => {
      if (fs.existsSync(pathToNextmdLastUpdate) === false || fs.existsSync(pathToNextmdBranch) === false) {
        return 'first time fetch';
      }

      const lastBranch = branchLastUpdate();
      if (lastBranch !== branch) {
        return ['was', lastBranch].join(' ');
      }

      if (elapsedSecondsSinceLastUpdate() > 5 * 60 * 1000) {
        // 5 minutes
        return 'last update was > 5 min ago';
      }

      return null;
    };

    const shouldUpdateGitRepoReason = shouldUpdateGitRepo();
    const logFromGit = [remoteUrl, `(Branch: ${branch})`].filter((e) => e).join(' ');
    if (shouldUpdateGitRepoReason) {
      consoleLogNextmd(
        ['cloning', logFromGit, debug ? `- ${shouldUpdateGitRepoReason}` : undefined].filter((e) => e).join(' '),
      );
      fs.rmSync(pathToContent, { recursive: true, force: true });
      await cmd(
        ['git', 'clone', branch ? `-b ${branch}` : undefined, '--depth 1', remoteUrl, pathToContent]
          .filter((e) => e)
          .join(' '),
      );
      fs.writeFileSync(pathToNextmdBranch, await cmd(`git -C ${pathToContent} rev-parse --abbrev-ref HEAD`));
      fs.writeFileSync(pathToNextmdLastUpdate, `${Date.now()}`);
    }

    consoleLogNextmd('creating page from', logFromGit);
  } else {
    consoleLogNextmd('creating page from', pathToContent);
  }

  return treeSync(pathToContent, absoluteToRelative);
};
