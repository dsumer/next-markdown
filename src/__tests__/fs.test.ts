import { TreeObject } from '../types';
import { exclude, getNextmdFromFilePath } from '../utils/fs';

describe('nextmd', () => {
  test('generate nextmd from path/index.abc', () => {
    expect(getNextmdFromFilePath('files/path/index.abc', 'files')).toEqual(['path']);
  });

  test('generate nextmd from path/hello.abc', () => {
    expect(getNextmdFromFilePath('files/path/hello.abc', 'files')).toEqual(['path', 'hello']);
  });

  test('generate nextmd from hello.abc', () => {
    expect(getNextmdFromFilePath('hello.abc', './')).toEqual(['hello']);
  });

  test('generate nextmd from _hello.abc', () => {
    expect(getNextmdFromFilePath('_hello.abc', './')).toEqual(['_hello']);
  });

  test('generate nextmd from 1970-01-01-helloworld.abc', () => {
    expect(getNextmdFromFilePath('1970-01-01-helloworld.abc', './')).toEqual(['helloworld']);
  });

  test('generate nextmd from path/1970-01-01-helloworld.abc', () => {
    expect(getNextmdFromFilePath('path/1970-01-01-helloworld.abc', './')).toEqual(['path', 'helloworld']);
  });
});

describe('tree object to exclude from generating a path', () => {
  const aFileWithExt = (extension: string): TreeObject => ({
    type: 'file',
    name: `hello.${extension}`,
    path: `path/hello.${extension}`,
  });

  const aFile = aFileWithExt('');

  test('include directories with children', () => {
    const gitDir: TreeObject = { type: 'dir', name: 'blog', path: 'path/to/blog', children: [aFile] };
    expect(exclude(gitDir)).toBe(false);
  });

  test('include md files', () => {
    expect(exclude(aFileWithExt('md'))).toBe(false);
  });

  test('include mdx files', () => {
    expect(exclude(aFileWithExt('mdx'))).toBe(false);
  });

  test('exclude empty directories', () => {
    const gitDir: TreeObject = { type: 'dir', name: 'blog', path: 'path/to/blog', children: [] };
    expect(exclude(gitDir)).toBe(true);
  });

  test('exclude .git directories', () => {
    const gitDir: TreeObject = { type: 'dir', name: '.git', path: 'path/to/.git', children: [aFile, aFile] };
    expect(exclude(gitDir)).toBe(true);
  });

  test('exclude files if not .md or .mdx', () => {
    expect(exclude(aFileWithExt('abc'))).toBe(true);
  });
});
