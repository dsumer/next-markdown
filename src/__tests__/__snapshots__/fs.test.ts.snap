// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree parsing 1`] = `
Array [
  Object {
    "name": "about.md",
    "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/about.md",
    "type": "file",
  },
  Object {
    "name": "b.md",
    "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/b.md",
    "type": "file",
  },
  Object {
    "children": Array [
      Object {
        "name": "index.md",
        "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/blog/index.md",
        "type": "file",
      },
      Object {
        "name": "post-1.md",
        "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/blog/post-1.md",
        "type": "file",
      },
      Object {
        "name": "post-2.mdx",
        "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/blog/post-2.mdx",
        "type": "file",
      },
      Object {
        "name": "post-3.md",
        "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/blog/post-3.md",
        "type": "file",
      },
    ],
    "name": "blog",
    "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/blog",
    "type": "dir",
  },
  Object {
    "children": Array [
      Object {
        "name": "c.md",
        "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/path/c.md",
        "type": "file",
      },
      Object {
        "children": Array [
          Object {
            "name": "e.mdx",
            "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/path/to/e.mdx",
            "type": "file",
          },
        ],
        "name": "to",
        "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/path/to",
        "type": "dir",
      },
    ],
    "name": "path",
    "path": "/Users/frouo/Programming/Web/next-markdown/src/__tests__/__filesystem__/path",
    "type": "dir",
  },
]
`;

exports[`flat files file and dirs with nested files 1`] = `
Array [
  Object {
    "name": "about",
    "path": "about",
    "type": "file",
  },
  Object {
    "name": "file1",
    "path": "docs/file1",
    "type": "file",
  },
  Object {
    "name": "file2",
    "path": "docs/file2",
    "type": "file",
  },
  Object {
    "name": "file1",
    "path": "blog/file1",
    "type": "file",
  },
]
`;

exports[`flat files file in a subfolder 1`] = `
Array [
  Object {
    "name": "file1",
    "path": "docs/subfolder/file1",
    "type": "file",
  },
]
`;

exports[`flat files only files 1`] = `
Array [
  Object {
    "name": "file1",
    "path": "file1",
    "type": "file",
  },
  Object {
    "name": "file2",
    "path": "file2",
    "type": "file",
  },
]
`;
